File: android/app/src/main/kotlin/co/tdevs/duplika/diagnostics/DiagnosticLogger.kt
package co.tdevs.duplika.diagnostics

import android.app.Application
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Process
import android.util.Log
import co.tdevs.duplika.BuildConfig
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.util.ArrayDeque
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

/**
 * The single sink for native diagnostics.
 *
 * Recording is cheap and never blocks the caller: build the event, push it into a
 * bounded ring, hand it to the live listener, and queue the disk write on a background
 * executor. The paths worth instrumenting — container install, guest launch, activity
 * start — are precisely the ones that must not get slower.
 *
 * Lives in **every** host process. `DuplikaApplication` is instantiated in Bcore's stub
 * processes too, so this object exists once per process, each with its own ring buffer
 * and its own log file. Only the main process has a Flutter engine, so only there does
 * [listener] get set; the other processes' events reach the console through the files
 * (see [DiagnosticStore]).
 */
object DiagnosticLogger {

    /** Recent events kept in memory, per process. */
    private const val RING_CAPACITY = 500

    /** Written together, so a burst of events costs one file write. */
    private const val FLUSH_THRESHOLD = 20

    private const val FLUSH_DELAY_MS = 1_500L

    private val sequence = AtomicLong(0)
    private val ring = ArrayDeque<DiagnosticEvent>(RING_CAPACITY)
    private val unwritten = ArrayList<DiagnosticEvent>(FLUSH_THRESHOLD)

    // Scheduled rather than plain: the delayed flush must not occupy the writer thread
    // while it waits, or an immediate batch queued behind it would be stuck for the
    // whole delay.
    private val writer = Executors.newSingleThreadScheduledExecutor { runnable ->
        Thread(runnable, "duplika-diagnostics").apply { isDaemon = true }
    }

    private val operationScope = ThreadLocal<OperationScope?>()

    @Volatile
    private var store: DiagnosticStore? = null

    @Volatile
    private var processName: String = "unknown"

    @Volatile
    private var appVersion: String? = null

    @Volatile
    private var appVersionCode: String? = null

    @Volatile
    private var deviceInfo: String = "${Build.MANUFACTURER} ${Build.MODEL}"

    @Volatile
    private var flushScheduled = false

    @Volatile
    private var idPrefix: String = "n0"

    /**
     * Where live events go when a Flutter engine is listening.
     *
     * Null in every process that has no Dart side, and null in the main process until
     * the console's channel attaches — which is why the ring buffer exists: the events
     * from before anyone was listening are replayed on attach.
     */
    @Volatile
    var listener: ((DiagnosticEvent) -> Unit)? = null

    /** True once [initialize] has resolved a log directory for this process. */
    val isInitialized: Boolean get() = store != null

    /**
     * Prepares logging for this process.
     *
     * Must be called from `Application.attachBaseContext`, **before** the virtualization
     * engine attaches. The engine rewrites this process's data directory for guest apps,
     * so resolving the log directory first is what keeps every process — host and stub
     * alike — writing into the host's real `filesDir` where the console can read them
     * back.
     *
     * Safe to call more than once; later calls are ignored.
     */
    fun initialize(context: Context) {
        if (store != null) return

        processName = resolveProcessName(context)
        idPrefix = "n${processName.substringAfterLast(':', "main").take(6)}" +
            "-${java.lang.Long.toString(System.currentTimeMillis(), 36)}"

        val directory = File(context.filesDir, DIRECTORY_NAME)
        store = DiagnosticStore(directory, sanitizeForFileName(processName))

        readVersion(context)
    }

    /** Fills in the app version. Kept separate so a failure here costs no logging. */
    private fun readVersion(context: Context) {
        try {
            val info = context.packageManager.getPackageInfo(context.packageName, 0)
            appVersion = info.versionName
            appVersionCode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                info.longVersionCode.toString()
            } else {
                @Suppress("DEPRECATION")
                info.versionCode.toString()
            }
        } catch (_: PackageManager.NameNotFoundException) {
            // A package manager that cannot find its own package happens inside a guest
            // process whose package visibility has been rewritten. Version stays null.
        } catch (_: Throwable) {
            // Version is a nicety; never let reading it break logging.
        }
    }

    fun currentProcessName(): String = processName

    fun currentAppVersion(): String? = appVersion

    fun currentAppVersionCode(): String? = appVersionCode

    fun buildType(): String = if (BuildConfig.DEBUG) "debug" else "release"

    fun retainedEventCount(): Int = synchronized(ring) { ring.size }

    fun occupiedBytes(): Long = store?.occupiedBytes() ?: 0L

    fun knownProcesses(): List<String> = store?.knownProcesses() ?: emptyList()

    /** Recent events from this process, oldest first. */
    fun recentEvents(): List<DiagnosticEvent> = synchronized(ring) { ring.toList() }

    /** The durable history across every process, oldest first. */
    fun history(limit: Int): List<DiagnosticEvent> {
        flushNow()
        return store?.readAll(limit) ?: recentEvents()
    }

    fun clear() {
        synchronized(ring) { ring.clear() }
        synchronized(unwritten) { unwritten.clear() }
        store?.clear()
    }

    // -------------------------------------------------------------------------
    // Correlation
    // -------------------------------------------------------------------------

    /** The operation in scope on this thread, if the Dart caller sent one. */
    fun currentOperation(): OperationScope? = operationScope.get()

    /**
     * Runs [block] with [id] as the ambient operation for this thread.
     *
     * Thread-local rather than a parameter on every call: the correlation id has to
     * reach code several layers down — the installer, the adapter, the engine's own
     * error paths — none of which should have to grow a diagnostics parameter.
     *
     * Restores the previous scope, so nesting works and an inner operation cannot
     * silently clear an outer one.
     */
    fun <T> withOperation(id: String?, name: String? = null, block: () -> T): T {
        if (id.isNullOrEmpty()) return block()
        val previous = operationScope.get()
        operationScope.set(OperationScope(id, name))
        return try {
            block()
        } finally {
            operationScope.set(previous)
        }
    }

    /** A correlation id captured on one thread and re-applied on another. */
    data class OperationScope(val id: String, val name: String?)

    // -------------------------------------------------------------------------
    // Recording
    // -------------------------------------------------------------------------

    fun debug(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        metadata: Map<String, String> = emptyMap(),
    ) = log(DiagLevel.DEBUG, source, category, message, metadata = metadata)

    fun info(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        packageName: String? = null,
        profileId: String? = null,
        virtualUserId: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ) = log(
        DiagLevel.INFO, source, category, message,
        packageName = packageName, profileId = profileId,
        virtualUserId = virtualUserId, metadata = metadata,
    )

    fun success(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        packageName: String? = null,
        profileId: String? = null,
        virtualUserId: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ) = log(
        DiagLevel.SUCCESS, source, category, message,
        packageName = packageName, profileId = profileId,
        virtualUserId = virtualUserId, metadata = metadata,
    )

    fun warning(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        error: Throwable? = null,
        packageName: String? = null,
        profileId: String? = null,
        virtualUserId: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ) = log(
        DiagLevel.WARNING, source, category, message, error,
        packageName = packageName, profileId = profileId,
        virtualUserId = virtualUserId, metadata = metadata,
    )

    fun error(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        error: Throwable? = null,
        packageName: String? = null,
        profileId: String? = null,
        virtualUserId: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ) = log(
        DiagLevel.ERROR, source, category, message, error,
        packageName = packageName, profileId = profileId,
        virtualUserId = virtualUserId, metadata = metadata,
    )

    fun fatal(
        source: DiagSource,
        category: DiagCategory,
        message: String,
        error: Throwable? = null,
        metadata: Map<String, String> = emptyMap(),
    ) = log(DiagLevel.FATAL, source, category, message, error, metadata = metadata)

    /**
     * The one method that creates events.
     *
     * [mirrorToLogcat] exists for [co.tdevs.duplika.native.Slog], which has already
     * written the line to Logcat by the time it forwards here; writing it again would
     * double every message in a `logcat` capture.
     */
    fun log(
        level: DiagLevel,
        source: DiagSource,
        category: DiagCategory,
        message: String,
        error: Throwable? = null,
        details: String? = null,
        operation: String? = null,
        operationName: String? = null,
        packageName: String? = null,
        profileId: String? = null,
        virtualUserId: Int? = null,
        metadata: Map<String, String> = emptyMap(),
        mirrorToLogcat: Boolean = true,
    ): DiagnosticEvent {
        val scope = operationScope.get()
        val next = sequence.getAndIncrement()

        val event = DiagnosticEvent(
            id = "$idPrefix-${java.lang.Long.toString(next, 36)}",
            timestampMillis = System.currentTimeMillis(),
            sequence = next,
            level = level,
            source = source,
            category = category,
            message = DiagRedactor.redactText(message).orEmpty(),
            operation = operation ?: scope?.id,
            operationName = operationName ?: scope?.name,
            details = DiagRedactor.redactText(details ?: error?.message),
            stackTrace = error?.let(::stackTraceOf),
            exceptionType = error?.javaClass?.name,
            packageName = packageName,
            profileId = profileId,
            processName = processName,
            virtualUserId = virtualUserId,
            thread = Thread.currentThread().name,
            buildType = buildType(),
            appVersion = appVersion,
            deviceInfo = deviceInfo,
            metadata = DiagRedactor.redactMetadata(
                if (error != null) metadata + ("errorMessage" to (error.message ?: "")) else metadata,
            ),
        )

        publish(event)
        if (mirrorToLogcat) {
            mirror(event, error)
        }
        return event
    }

    /** Sends the in-memory ring to a listener that has just attached. */
    fun replayTo(target: (DiagnosticEvent) -> Unit) {
        recentEvents().forEach(target)
    }

    private fun publish(event: DiagnosticEvent) {
        synchronized(ring) {
            if (ring.size >= RING_CAPACITY) ring.removeFirst()
            ring.addLast(event)
        }

        listener?.let { target ->
            try {
                target(event)
            } catch (_: Throwable) {
                // A detached channel throws on send. Losing the live copy is fine; the
                // event is already in the ring and queued for disk.
            }
        }

        // DEBUG stays in memory. It is the noisiest level and the least useful after
        // the fact, and the file budget is better spent on everything else.
        if (event.level == DiagLevel.DEBUG) return

        val batch: List<DiagnosticEvent>? = synchronized(unwritten) {
            unwritten.add(event)
            if (unwritten.size >= FLUSH_THRESHOLD) {
                val copy = ArrayList(unwritten)
                unwritten.clear()
                copy
            } else {
                null
            }
        }

        if (batch != null) {
            submit { store?.append(batch) }
        } else {
            scheduleFlush()
        }
    }

    private fun scheduleFlush() {
        if (flushScheduled) return
        flushScheduled = true
        try {
            writer.schedule(
                {
                    flushScheduled = false
                    try {
                        flushNow()
                    } catch (_: Throwable) {
                        // See DiagnosticStore.append.
                    }
                },
                FLUSH_DELAY_MS,
                TimeUnit.MILLISECONDS,
            )
        } catch (_: RejectedExecutionException) {
            flushScheduled = false
        }
    }

    /** Writes anything queued. Called before any read, so a history is never stale. */
    fun flushNow() {
        val batch = synchronized(unwritten) {
            if (unwritten.isEmpty()) return
            val copy = ArrayList(unwritten)
            unwritten.clear()
            copy
        }
        store?.append(batch)
    }

    private fun submit(work: () -> Unit) {
        try {
            writer.execute {
                try {
                    work()
                } catch (_: Throwable) {
                    // Swallowed on purpose: see DiagnosticStore.append.
                }
            }
        } catch (_: RejectedExecutionException) {
            // Process is going down. The event is still in the ring for anyone who asks.
        }
    }

    /**
     * Keeps Logcat working exactly as before diagnostics existed.
     *
     * Everything Duplika used to print still prints, under the same `Duplika.*` tags,
     * so an existing `adb logcat` workflow is unaffected.
     */
    private fun mirror(event: DiagnosticEvent, error: Throwable?) {
        val tag = "Duplika.${event.source.wire}"
        val line = if (event.operation != null) {
            "[${event.operation}] ${event.message}"
        } else {
            event.message
        }
        when (event.level) {
            DiagLevel.DEBUG -> Log.d(tag, line)
            DiagLevel.INFO, DiagLevel.SUCCESS -> Log.i(tag, line)
            DiagLevel.WARNING -> Log.w(tag, line, error)
            DiagLevel.ERROR, DiagLevel.FATAL -> Log.e(tag, line, error)
        }
    }

    private fun stackTraceOf(error: Throwable): String {
        val writer = StringWriter()
        error.printStackTrace(PrintWriter(writer))
        val text = writer.toString().trimEnd()
        // A deep Kotlin/AOSP trace can run to hundreds of lines. Keeping all of them
        // would let a handful of exceptions consume the whole size-bounded log.
        val lines = text.lines()
        return if (lines.size <= MAX_STACK_LINES) {
            text
        } else {
            lines.take(MAX_STACK_LINES).joinToString("\n") +
                "\n\t... ${lines.size - MAX_STACK_LINES} more lines"
        }
    }

    /**
     * The process this code is running in.
     *
     * `Application.getProcessName()` is API 28+; below that the only reliable answer is
     * the running-app-process list, and even that can come back empty on some OEM
     * builds, so an unknown process name is reported rather than guessed.
     */
    private fun resolveProcessName(context: Context): String {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            return runCatching { Application.getProcessName() }.getOrNull()
                ?: context.packageName
        }
        val pid = Process.myPid()
        return runCatching {
            val manager = context.getSystemService(Context.ACTIVITY_SERVICE)
                as? android.app.ActivityManager
            manager?.runningAppProcesses
                ?.firstOrNull { it.pid == pid }
                ?.processName
        }.getOrNull() ?: context.packageName
    }

    private fun sanitizeForFileName(value: String): String =
        value.replace(Regex("[^A-Za-z0-9_.-]"), "_").take(48)

    private const val DIRECTORY_NAME = "diagnostics"
    private const val MAX_STACK_LINES = 60
}

File: lib/core/diagnostics/diagnostic_logger.dart
import 'dart:async';
import 'dart:developer' as developer;

import 'package:flutter/foundation.dart';

import 'diagnostic_buffer.dart';
import 'diagnostic_event.dart';
import 'diagnostic_operation.dart';
import 'diagnostic_redactor.dart';
import 'diagnostic_store.dart';

/// The single sink every Dart-side diagnostic goes through.
///
/// Recording is synchronous and cheap by design — build the event, redact it, push it
/// into the ring buffer, hand it to the broadcast stream — and everything expensive
/// (disk) is batched behind a timer. Nothing on a caller's path awaits I/O, because the
/// paths worth instrumenting are exactly the ones that must not get slower: APK import,
/// container install, guest launch.
///
/// Failure of the logger is never allowed to become the failure under investigation:
/// persistence errors are swallowed, and the buffer is bounded.
class DiagnosticLogger {
  DiagnosticLogger({
    DiagnosticBuffer? buffer,
    DiagnosticEventStore? store,
    this.flushInterval = const Duration(seconds: 2),
    this.flushThreshold = 25,
  })  : buffer = buffer ?? DiagnosticBuffer(),
        _store = store ?? const NoopDiagnosticEventStore();

  /// The app-wide logger. Replaceable so tests can install one with a fake store.
  static DiagnosticLogger get instance => _instance ??= DiagnosticLogger();

  static DiagnosticLogger? _instance;

  @visibleForTesting
  static set instance(DiagnosticLogger logger) => _instance = logger;

  /// Installs the production logger: bounded buffer plus the on-disk store.
  ///
  /// Safe to call more than once; the first call wins, so a test that installed its
  /// own logger is not overwritten by app start-up.
  static DiagnosticLogger bootstrap() =>
      _instance ??= DiagnosticLogger(store: FileDiagnosticEventStore());

  final DiagnosticBuffer buffer;
  final DiagnosticEventStore _store;

  /// How long a batch may sit in memory before being written. Long enough that a burst
  /// of events costs one write, short enough that a crash loses at most this much.
  final Duration flushInterval;

  /// A batch this large is written immediately rather than waiting out the timer.
  final int flushThreshold;

  final StreamController<DiagnosticEvent> _events =
      StreamController<DiagnosticEvent>.broadcast();

  final List<DiagnosticEvent> _unwritten = <DiagnosticEvent>[];
  Timer? _flushTimer;

  int _sequence = 0;
  late final String _idPrefix = _makeIdPrefix();

  String? _appVersion;
  String? _deviceInfo;
  String? _processName;

  /// Live events, newest as they happen. Broadcast, so the console can attach and
  /// detach without affecting recording.
  Stream<DiagnosticEvent> get stream => _events.stream;

  /// `debug`, `profile` or `release`. Stamped on every event so a report from a release
  /// build cannot be mistaken for a debug one.
  static String get buildType {
    if (kDebugMode) {
      return 'debug';
    }
    if (kProfileMode) {
      return 'profile';
    }
    return 'release';
  }

  /// Fills in the fields that need the platform to answer first.
  ///
  /// Called once system information has been read from native. Events recorded before
  /// that simply have no version or device — an absent field is honest, a guessed one
  /// is not.
  void describeEnvironment({
    String? appVersion,
    String? deviceInfo,
    String? processName,
  }) {
    _appVersion = appVersion ?? _appVersion;
    _deviceInfo = deviceInfo ?? _deviceInfo;
    _processName = processName ?? _processName;
  }

  DiagnosticEvent record({
    required DiagLevel level,
    required DiagnosticSource source,
    required DiagnosticCategory category,
    required String message,
    String? details,
    Object? error,
    StackTrace? stackTrace,
    String? operation,
    String? operationName,
    String? packageName,
    String? profileId,
    int? virtualUserId,
    Map<String, String>? metadata,
  }) {
    final DiagnosticOperation? ambient = DiagnosticOperation.current;

    final DiagnosticEvent event = DiagnosticEvent(
      id: '$_idPrefix-${_sequence.toRadixString(36)}',
      sequence: _sequence++,
      timestamp: DateTime.now(),
      level: level,
      source: source,
      category: category,
      message: DiagnosticRedactor.redactText(message),
      details: details == null ? null : DiagnosticRedactor.redactText(details),
      exceptionType: error?.runtimeType.toString(),
      // The exception's own text is part of the failure, so it is preserved rather than
      // summarised — redaction is the only thing applied to it.
      stackTrace: stackTrace == null ? null : _trimStack(stackTrace),
      operation: operation ?? ambient?.id,
      operationName: operationName ?? ambient?.name,
      packageName: packageName ?? ambient?.packageName,
      profileId: profileId ?? ambient?.profileId,
      virtualUserId: virtualUserId,
      processName: _processName,
      thread: 'dart:main',
      buildType: buildType,
      appVersion: _appVersion,
      deviceInfo: _deviceInfo,
      metadata: _buildMetadata(error, metadata),
    );

    _publish(event);
    _mirrorToConsole(event, error, stackTrace);
    return event;
  }

  /// Takes an event that was built elsewhere — the Kotlin logger, or a stored history
  /// being replayed — without re-stamping its identity.
  ///
  /// Not persisted by default. Native events already have a durable copy written by
  /// the Kotlin store, and writing them again here would spend the Dart log's byte
  /// budget on a second copy of events the export already reads from native.
  void recordExternal(DiagnosticEvent event, {bool persist = false}) {
    _publish(event, persist: persist);
  }

  void debug(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.debug,
        source: source,
        category: category,
        message: message,
        metadata: metadata,
      );

  void info(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    String? packageName,
    String? profileId,
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.info,
        source: source,
        category: category,
        message: message,
        packageName: packageName,
        profileId: profileId,
        metadata: metadata,
      );

  void success(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    String? packageName,
    String? profileId,
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.success,
        source: source,
        category: category,
        message: message,
        packageName: packageName,
        profileId: profileId,
        metadata: metadata,
      );

  void warning(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    Object? error,
    StackTrace? stackTrace,
    String? packageName,
    String? profileId,
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.warning,
        source: source,
        category: category,
        message: message,
        error: error,
        stackTrace: stackTrace,
        packageName: packageName,
        profileId: profileId,
        metadata: metadata,
      );

  void error(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    Object? error,
    StackTrace? stackTrace,
    String? packageName,
    String? profileId,
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.error,
        source: source,
        category: category,
        message: message,
        error: error,
        stackTrace: stackTrace,
        packageName: packageName,
        profileId: profileId,
        metadata: metadata,
      );

  void fatal(
    DiagnosticSource source,
    DiagnosticCategory category,
    String message, {
    Object? error,
    StackTrace? stackTrace,
    Map<String, String>? metadata,
  }) =>
      record(
        level: DiagLevel.fatal,
        source: source,
        category: category,
        message: message,
        error: error,
        stackTrace: stackTrace,
        metadata: metadata,
      );

  /// Everything the ring buffer still holds, oldest first.
  List<DiagnosticEvent> recentEvents() => buffer.toList();

  /// The persisted history, oldest first, newest-biased when truncated.
  Future<List<DiagnosticEvent>> history({int limit = 0}) async {
    await flush();
    return _store.read(limit: limit);
  }

  Future<int> persistedBytes() => _store.occupiedBytes();

  /// Writes anything still queued. Called before reading history or exporting, so a
  /// report can never be missing the event that prompted it.
  ///
  /// A store failure is absorbed here rather than left to the caller. Most flushes are
  /// fire-and-forget from [_publish], so a store that threw would surface as an
  /// unhandled async error — the logging system becoming the incident, which is the one
  /// thing it must never do.
  Future<void> flush() {
    _flushTimer?.cancel();
    _flushTimer = null;

    final List<DiagnosticEvent> batch = List<DiagnosticEvent>.of(_unwritten);
    _unwritten.clear();
    return _store.append(batch).catchError((Object _) {});
  }

  /// Drops both the recent buffer and the persisted history.
  Future<void> clear() async {
    _flushTimer?.cancel();
    _flushTimer = null;
    _unwritten.clear();
    buffer.clear();
    await _store.clear();
  }

  void _publish(DiagnosticEvent event, {bool persist = true}) {
    buffer.add(event);
    if (_events.hasListener) {
      _events.add(event);
    }

    // Debug-level events are the noisiest and the least useful after the fact, so they
    // stay in memory only. Everything a bug report would want is persisted.
    if (persist && _store.isPersistent && event.level != DiagLevel.debug) {
      _unwritten.add(event);
      if (_unwritten.length >= flushThreshold) {
        flush();
      } else {
        _flushTimer ??= Timer(flushInterval, flush);
      }
    }
  }

  Map<String, String> _buildMetadata(Object? error, Map<String, String>? extra) {
    final Map<String, String> metadata = <String, String>{
      ...?extra,
      'error': ?error?.toString(),
    };
    return DiagnosticRedactor.redactMetadata(metadata);
  }

  /// Keeps the frames that identify the failure and drops the rest.
  ///
  /// A full Flutter stack is ~100 frames of framework internals; storing all of them
  /// would let a handful of errors consume the whole size-bounded log.
  static String _trimStack(StackTrace stackTrace) {
    const int maxFrames = 40;
    final List<String> frames = stackTrace.toString().split('\n');
    if (frames.length <= maxFrames) {
      return stackTrace.toString().trimRight();
    }
    return <String>[
      ...frames.take(maxFrames),
      '... ${frames.length - maxFrames} more frames',
    ].join('\n');
  }

  /// Keeps `dart:developer` output working exactly as it did before diagnostics
  /// existed, so nothing that used to appear in the IDE console stops appearing.
  ///
  /// In release only problems are mirrored: the console call is not free, and there is
  /// no attached observatory to read the rest.
  void _mirrorToConsole(DiagnosticEvent event, Object? error, StackTrace? stackTrace) {
    if (!kDebugMode && !event.isProblem) {
      return;
    }
    developer.log(
      event.message,
      name: 'Duplika.${event.source.wire}',
      level: _developerLevel(event.level),
      error: error,
      stackTrace: stackTrace,
    );
  }

  static int _developerLevel(DiagLevel level) => switch (level) {
        DiagLevel.debug => 500,
        DiagLevel.info => 800,
        DiagLevel.success => 800,
        DiagLevel.warning => 900,
        DiagLevel.error => 1000,
        DiagLevel.fatal => 1200,
      };

  /// A short per-run token, so event ids from two runs (or from a native process) can
  /// never collide even though the counter restarts at zero.
  static String _makeIdPrefix() {
    final int micros = DateTime.now().microsecondsSinceEpoch;
    return 'd${micros.toRadixString(36)}';
  }
}
